### ISSUE TITLE
Core: Python API has no `__repr__` which makes debugging very user unfriendly  

### ISSUE THREAD

**Issue by AIRCAP** [FreeCAD/FreeCAD#17059]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

As discussed in #development:

> @benj5378: @CorvusCorax idk if an issue already exists... but please open an issue on GitHub requesting implementation of `__repr__` across objects in all of FreeCAD. It's actually a very big issue, making it very difficult to use Python Console.
> 
>[https://docs.python.org/3/reference/datamodel.html#object.\_\_repr\_\_](https://docs.python.org/3/reference/datamodel.html#object.__repr__)
> 
> Like even this
> 
> >>> Gui.ActiveDocument
>` <GUI Document object at 0x5791140ea400>`
> 
> If `__repr__` was implemented, it would be
> 
> >>> Gui.ActiveDocument
> `<GUI Document MyNewTable>`
> 
> Really bad user experience
> 

We also found out that `Gui.Control.activeDialog()` only returns true or false instead of a python representation of the dialog to interact . Instead there is `Gui.Control.activeTaskDialog()` but it has no `__repr__` so you can't tell which one.

This issue could act as a meta-issue for general improvements and housecleaning of the Python API in case more are found - in any way this is post 1.0 work and not super urgent.

### Full version info

```shell
all current versions
```

### Subproject(s) affected?

Other (specify in description)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by benj5378** [comment]
`Gui.Control.activeTaskDialog()` returns `Task Dialog` as `__repr__` for the class. It would be much better if it returned `<Task Dialog Sketcher Edit>` for instance

**Comment by benj5378** [comment]
@tritao, is this something for you? Preferably, compiltion/tests should fail if no implemented __repr__

**Comment by benj5378** [comment]
Still relevant. @luzpaz can you add easy/good first issue label?

**Comment by luzpaz** [comment]
Hi @pyro9, any interest in taking this on? Since your FreeCAD tutorials heavily utilize the python console these days... this could really soup up the python UX ? No pressure, was just wondering if you'd be interested. 😅

**Comment by tritao** [comment]
cc @mnesarco as well, who has been looking into things in this area.

**Comment by mnesarco** [comment]
> cc [@mnesarco](https://github.com/mnesarco) as well, who has been looking into things in this area.

Yeah another area to improve..... 1.2 will have some progress with python integration to improve QOL for devs... there are so many areas that need some care.

**Comment by luzpaz** [comment]
Can someone show an example of the correct use of `__repr__` per this ticket in the FreeCAD source code ? 

We should also create a link on the developer handbook repo for devs to start incorporating it

Edit: 
* https://github.com/FreeCAD/DevelopersHandbook/issues/139

**Comment by pyro9** [comment]
@luzpaz Yes, I'd like to take a look at this. Both for the Python representation of some core objects and to get FeaturePython to allow and use a __repr__ method from the Python side.

**Comment by luzpaz** [comment]
@pyro9 thank you so much!

Question: so am I correct to think that if a python API function doesn't have an assigned  `__repr__` we will now see that in the python console output ?

**Comment by pyro9** [comment]
@luzpaz So far I've just made a few __repr__ more informative. I'm still peeling the onion on this one.

**Comment by chennes** [comment]
(this shouldn't have been auto-closed, this is a pretty large work-in-progress... reopened, but let me know if there's a more useful way to track this work, @pyro9)

**Comment by pyro9** [comment]
@chennes , I'm not really sure. This issue will do for tracking for now.

**Comment by luzpaz** [comment]
is there a script we can conjure up that scans the python API for the presence of `__repr__`  and reports the where it is missing? We can then generate a markdown checkmark list to track the progress of this effort ?

**Comment by mnesarco** [comment]
> is there a script we can conjure up that scans the python API for the presence of `__repr__` and reports the where it is missing? We can then generate a markdown checkmark list to track the progress of this effort ?

AFAIK it is present in all bindings, but probably with the minimal implementation.

**Comment by pyro9** [comment]
@luzpaz, @mnesarco In some cases, the base class provides repr for all of the children, but it provides the same generic repr. For example all PartFeatures call themselves <Part::Feature> (see #24429) Once I peeled that back, All Part::FeaturePython reported only <Part::FeaturePython> (patch forthcoming to include the Python class' repr)

Also interestingly if the Feature is attachable, the repr you get from just the object on the console comes from PartFeaturePyImp.cpp but if you explicitly call obj.__repr__() it comes from AttachExtensionPyImp.cpp instead (so for now I had to duplicate my changes to PartFeaturePyImp there).

But technically, they all had  __repr__. I may need to go workbench by workbench.
